File: internal/lore/world.go
package lore

import (
	"errors"
	"fmt"
	"strings"
)

// World holds the parsed entity graph and reference index. Every element of
// Entities has a non-empty Type after Merge returns — untyped header lines are
// either attached to an existing typed entity or dropped.
type World struct {
	Entities   []Entity
	References map[string][]Reference
	// Match is a pre-lowered lookup cache built by Merge. It stays in lock
	// step with Entities — rebuilt whenever Merge runs — so callers that
	// scan text for entity mentions don't need to re-lowercase on every
	// iteration of their hot loop.
	Match *MatchIndex
	// Files retains the raw content of every parsed file so Search can scan
	// free text alongside entity descriptions — narrative prose between
	// definitions is part of the searchable surface per docs/design.md.
	Files []FileSource
	// FileIssues holds diagnostics not tied to a specific entity — most
	// notably untyped colon-line typos: a `Sildar Hallwinder:` block that
	// matches no known entity but is close to one. Phase 2 of Merge would
	// silently drop these (they look like prose); FileIssues lets the LSP
	// and `check` surface them so authors notice the gap.
	FileIssues []StateIssue
}

// FileSource is one parsed file's path and raw content, kept on the World so
// full-text search can scan prose outside any entity definition.
type FileSource struct {
	Path    string
	Content string
}

// NewWorld creates an empty World.
func NewWorld() *World {
	return &World{
		References: make(map[string][]Reference),
	}
}

// AddReference records a reference to the named entity.
func (w *World) AddReference(entityName string, ref Reference) {
	w.References[entityName] = append(w.References[entityName], ref)
}

var ErrNotFound = errors.New("entity not found")

// AmbiguousError is returned when a bare name matches multiple entities.
type AmbiguousError struct {
	Matches []*Entity
}

func (e *AmbiguousError) Error() string {
	names := make([]string, len(e.Matches))
	for i, m := range e.Matches {
		if m.Type != "" {
			names[i] = fmt.Sprintf("%s (%s)", m.Name, m.Type)
		} else {
			names[i] = m.Name
		}
	}
	return fmt.Sprintf("ambiguous: %s", strings.Join(names, ", "))
}

// FindEntity looks up an entity by name or alias, supporting disambiguation syntax.
func (w *World) FindEntity(name string) (*Entity, error) {
	// Try disambiguation syntax first: "Barovia (town)".
	if disambig, ok := ParseDisambiguation(name); ok {
		for i := range w.Entities {
			ent := &w.Entities[i]
			if ent.Type == "" || !strings.EqualFold(ent.Type, disambig.Type) {
				continue
			}
			if strings.EqualFold(ent.Name, disambig.Name) || ent.NameMatchesAlias(disambig.Name) {
				return ent, nil
			}
		}
		return nil, ErrNotFound
	}

	// Collect all matches to detect ambiguity.
	var matches []*Entity
	for i := range w.Entities {
		ent := &w.Entities[i]
		if strings.EqualFold(ent.Name, name) || ent.NameMatchesAlias(name) {
			matches = append(matches, ent)
		}
	}

	switch len(matches) {
	case 0:
		return nil, ErrNotFound
	case 1:
		return matches[0], nil
	default:
		return nil, &AmbiguousError{Matches: matches}
	}
}

// GetReferences returns all references to an entity (case-insensitive name lookup),
// excluding self-references (where the entity's own definition mentions itself).
func (w *World) GetReferences(name string) []Reference {
	var raw []Reference
	if refs, ok := w.References[name]; ok {
		raw = refs
	} else {
		for key, refs := range w.References {
			if strings.EqualFold(key, name) {
				raw = refs
				break
			}
		}
	}
	if len(raw) == 0 {
		return nil
	}

	// Find the entity so we can match against its canonical name and aliases.
	ent, err := w.FindEntity(name)
	if err != nil {
		return raw
	}

	filtered := make([]Reference, 0, len(raw))
	for _, ref := range raw {
		if ref.SourceEntity != "" && isSameEntity(ent, ref.SourceEntity, ref.SourceType) {
			continue
		}
		filtered = append(filtered, ref)
	}
	return filtered
}

// isSameEntity reports whether (sourceName, sourceType) refers to the
// same entity. When sourceType is non-empty it must also match — so
// cross-references between two entities sharing a bare name (e.g.
// "Barovia (town)" mentioning "Barovia (country)") aren't mistaken for
// self-references.
func isSameEntity(ent *Entity, sourceName, sourceType string) bool {
	if !strings.EqualFold(ent.Name, sourceName) && !ent.NameMatchesAlias(sourceName) {
		return false
	}
	if sourceType == "" || ent.Type == "" {
		return true
	}
	return strings.EqualFold(ent.Type, sourceType)
}

// Search finds lines containing the query text (case-insensitive) across
// every parsed file. Both entity descriptions and free text prose are
// scanned, since narrative outside any definition is part of the
// searchable surface (docs/design.md, docs/format.md). Returns one result
// per matching line, sorted by file then line.
func (w *World) Search(query string) []SearchResult {
	if query == "" {
		return nil
	}
	var results []SearchResult
	for _, f := range w.Files {
		lines := strings.Split(f.Content, "\n")
		for i, line := range lines {
			if !ContainsIgnoreCase(line, query) {
				continue
			}
			results = append(results, SearchResult{
				File:    f.Path,
				Line:    i + 1,
				Context: strings.TrimSpace(line),
			})
		}
	}
	return results
}

// Check reports issues such as undefined references and entity state problems.
func (w *World) Check() []Issue {
	var issues []Issue
	for _, ent := range w.Entities {
		for _, si := range ent.StateIssues {
			issues = append(issues, Issue{
				File:    si.Span.File,
				Line:    si.Span.Line,
				Message: si.Message,
			})
		}
	}
	for _, fi := range w.FileIssues {
		issues = append(issues, Issue{
			File:    fi.Span.File,
			Line:    fi.Span.Line,
			Message: fi.Message,
		})
	}
	return issues
}

File: legacy/src/world.zig
const std = @import("std");
const entity = @import("entity.zig");

const Entity = entity.Entity;
const Description = entity.Description;
const Reference = entity.Reference;
const SearchResult = entity.SearchResult;
const Issue = entity.Issue;

pub const World = struct {
    /// Arena owns all string data allocated during parsing. Heap-allocated
    /// to avoid invalidation when World is moved/returned.
    arena: *std.heap.ArenaAllocator,
    allocator: std.mem.Allocator,
    entities: std.ArrayList(Entity),
    references: std.StringHashMap(std.ArrayList(Reference)),

    pub fn deinit(self: *World) void {
        for (self.entities.items) |*e| {
            e.aliases.deinit(self.allocator);
            e.descriptions.deinit(self.allocator);
        }
        self.entities.deinit(self.allocator);
        var it = self.references.iterator();
        while (it.next()) |entry| {
            entry.value_ptr.deinit(self.allocator);
        }
        self.references.deinit();
        self.arena.deinit();
        self.allocator.destroy(self.arena);
    }

    const max_ambiguous = 16;

    pub const FindResult = union(enum) {
        found: *const Entity,
        not_found,
        ambiguous: struct {
            entities: [max_ambiguous]*const Entity,
            count: usize,

            pub fn items(self: *const @This()) []const *const Entity {
                return self.entities[0..self.count];
            }
        },

        /// Returns the entity if found, null otherwise.
        pub fn get(self: FindResult) ?*const Entity {
            return switch (self) {
                .found => |e| e,
                else => null,
            };
        }
    };

    pub fn findEntity(self: *const World, name: []const u8) FindResult {
        // Support disambiguation syntax: "Barovia (town)"
        if (entity.parseDisambiguation(name)) |disambig| {
            for (self.entities.items) |*e| {
                if (e.entity_type) |et| {
                    if (!std.ascii.eqlIgnoreCase(et, disambig.entity_type)) continue;
                    if (std.ascii.eqlIgnoreCase(e.name, disambig.name)) return .{ .found = e };
                    if (entity.nameMatchesAlias(e, disambig.name)) return .{ .found = e };
                }
            }
            return .not_found;
        }

        // Collect all matches to detect ambiguity
        var matches: [max_ambiguous]*const Entity = undefined;
        var match_count: usize = 0;

        for (self.entities.items) |*e| {
            if (std.ascii.eqlIgnoreCase(e.name, name) or entity.nameMatchesAlias(e, name)) {
                if (match_count < matches.len) {
                    matches[match_count] = e;
                    match_count += 1;
                }
            }
        }

        return switch (match_count) {
            0 => .not_found,
            1 => .{ .found = matches[0] },
            else => .{ .ambiguous = .{ .entities = matches, .count = match_count } },
        };
    }

    pub fn getReferences(self: *const World, name: []const u8) std.ArrayList(Reference) {
        if (self.references.get(name)) |refs| return refs;
        var it = self.references.iterator();
        while (it.next()) |entry| {
            if (std.ascii.eqlIgnoreCase(entry.key_ptr.*, name)) return entry.value_ptr.*;
        }
        var empty: std.ArrayList(Reference) = .empty;
        _ = &empty;
        return empty;
    }

    pub fn search(self: *const World, query: []const u8) std.ArrayList(SearchResult) {
        var results: std.ArrayList(SearchResult) = .empty;
        for (self.entities.items) |e| {
            for (e.descriptions.items) |desc| {
                if (entity.containsIgnoreCase(desc.text, query)) {
                    results.append(self.allocator, .{
                        .file = desc.file,
                        .line = desc.line,
                        .context = desc.text,
                    }) catch {};
                }
            }
        }
        return results;
    }

    pub fn check(self: *const World) std.ArrayList(Issue) {
        _ = self;
        var empty: std.ArrayList(Issue) = .empty;
        _ = &empty;
        return empty;
    }
};
